openedx/tests/xblock_integration/xblock_testcase.py

```diff
@@ -0,0 +1,387 @@
+        '''
+        def capture_score(user_id, usage_key, score, max_score):
+            '''
+            Hot-patch which stores scores in a local array instead of the
+            database.
+
+            Note that to make this generic, we'd need to do both.
+            '''
+            self.scores.append({'student': user_id,
+                                'usage': usage_key,
+                                'score': score,
+                                'max_score': max_score})
+
+        super(GradePublishTestMixin, self).setUp()
+
+        self.scores = []
+        patcher = mock.patch("courseware.module_render.set_score", capture_score)
+        patcher.start()
+        self.addCleanup(patcher.stop)
+
+    def assert_grade(self, grade):
+        '''
+        Confirm that the last grade set was equal to grade.
+
+        HACK: In the future, this should take a user ID and a block url_name.
+        '''
+        self.assertEqual(grade, self.scores[-1]['score'])
+
+
+class XBlockScenarioTestCaseMixin(object):
+    '''
+    This allows us to have test cases defined in JSON today, and in OLX
+    someday.
+    '''
+    @classmethod
+    def setUpClass(cls):
+        """
+        Create a page with two of the XBlock on it
+        """
+        super(XBlockScenarioTestCaseMixin, cls).setUpClass()
+
+        cls.course = CourseFactory.create(
+            display_name='XBlock_Test_Course'
+        )
+        cls.scenario_urls = {}
+        with cls.store.bulk_operations(cls.course.id, emit_signals=False):
+            for chapter_config in cls.test_configuration:
+                chapter = ItemFactory.create(
+                    parent=cls.course,
+                    display_name="ch_" + chapter_config['urlname'],
+                    category='chapter'
+                )
+                section = ItemFactory.create(
+                    parent=chapter,
+                    display_name="sec_" + chapter_config['urlname'],
+                    category='sequential'
+                )
+                unit = ItemFactory.create(
+                    parent=section,
+                    display_name='New Unit',
+                    category='vertical'
+                )
+                for xblock_config in chapter_config['xblocks']:
+                    ItemFactory.create(
+                        parent=unit,
+                        category=xblock_config['blocktype'],
+                        display_name=xblock_config['urlname']
+                    )
+
+                scenario_url = unicode(reverse(
+                    'courseware_section',
+                    kwargs={
+                        'course_id': unicode(cls.course.id),
+                        'chapter': "ch_" + chapter_config['urlname'],
+                        'section': "sec_" + chapter_config['urlname']
+                    }
+                ))
+
+                cls.scenario_urls[chapter_config['urlname']] = scenario_url
```

Comment:
Check out https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/tests/xml/factories.py for one place where we actually dynamically load in OLX. We can probably do a similar thing here (in a later PR) to allow the tests to use OLX definitions.